### PR TITLE
Beginnings of an evaluation module.

### PR DESCRIPTION
**Pull request by akleeman**
Includes:
 - A refactor of solution.py to work with multiple filters.
 - Consolidation of tools
 - Quantitative evaluation as well as visual evaluation in compare.py.

@swift-nav/estimation

### CODE REVIEW

File: gnss_analysis/evaluation/metrics.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import pandas as pd
+
+
+def ecef_from_info(info_dict):
+  """
+  Given an info dictionary this returns the ecef position of the
+  receiver.  If the position attributes don't exist this will
+  return nans.
+  """
+  return np.array([info_dict.get('x', np.nan),
```
Comment (benjamin0) [inline gnss_analysis/evaluation/metrics.py]: maybe fail more explicitly if position doesn't exist?

File: gnss_analysis/evaluation/metrics.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  first_time, fix_time = time_to_convergence(baseline_errors, threshold)
+  if first_time == np.datetime64('NaT'):
+    return np.array([np.nan])
+  # then exhaust the rest of the errors and compute summary statistics.
+  return np.array([np.asscalar(x['baseline_error'].values)
+                  for x in baseline_errors
+                  if get_time(x) > fix_time])
+
+
+def print_evaluations(soln_sets, filter_name):
```
Comment (benjamin0) [inline gnss_analysis/evaluation/metrics.py]: add threshold as a parameter here?

File: gnss_analysis/evaluation/visual.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        plt.show()
+        plt.pause(0.001)
+      else:
+        # otherwise update the plot by first resetting the data of the line plots
+        for line, vals in zip(lines, metric_buffer.values.T):
+          line.set_data(epoch_buffer,
+                        vals)
+        # then (possibly) adjusting the xlimits of the plot
+        if epoch_buffer[-1] > ax.get_xlim()[1]:
+          ax.set_xlim([0, 1.61 * epoch_buffer[-1]])
```
Comment (benjamin0) [inline gnss_analysis/evaluation/visual.py]: golden ratio?

File: gnss_analysis/filters/__init__.py [hunk 4]
```diff
@@ -1,17 +1,17 @@
...
-lookup = frozendict({'L1-static': functools.partial(StaticKalmanFilter,
+# the lookup table should not include an operator symbols (such as '-')
+lookup = frozendict({'L1_static': functools.partial(StaticKalmanFilter,
                                                     single_band=True),
-                     'multiband-static': functools.partial(StaticKalmanFilter,
+                     'multiband_static': functools.partial(StaticKalmanFilter,
                                                            single_band=False),
-                     'static': StaticKalmanFilter,
-                     'dynamic': DynamicKalmanFilter,
+                     'L1_dynamic': functools.partial(DynamicKalmanFilter,
```
Comment (benjamin0) [inline gnss_analysis/filters/__init__.py]: should we add multiband dynamic (and in the future every combination)?

File: gnss_analysis/filters/common.py [hunk 5]
```diff
@@ -1,8 +1,18 @@
...
-from gnss_analysis import dgnss, ephemeris, propagate, solution
+
+def can_compute_position(obs):
+  """
+  Convenience function that returns true if the state contains
+  enough information to compute a position.
+  """
+  # make sure we have enough rover observations
+  if obs.empty or obs.shape[0] < 4:
+    logging.info("Waiting for 4 satellites, only %d found so far."
```
Comment (benjamin0) [inline gnss_analysis/filters/common.py]: maybe make this a warning? when there was the nav parsing error it looked like nothing was happening even though it couldn't find 4 satellites

**Comment by benjamin0** [comment]
could you also update the readme to show an example of running the eval module?

File: gnss_analysis/tools/compare.py [hunk 6]
```diff
@@ -1,170 +1,81 @@
...
+  else:
+    soln_sets = list(soln_sets)
+    for fname in args.filters.keys():
+      metrics.print_evaluations(soln_sets, fname)
 
 
 
 def create_parser(parser):
-  parser.add_argument('input', type=argparse.FileType('r'),
+  parser.add_argument('input',
```
Comment (benjamin0) [inline gnss_analysis/tools/compare.py]: could we add an option to just plot at the end and not in real time?

**Comment by benjamin0** [comment]
could you also update the readme to show an example of running the eval module? (if the current readme section on compare.py is out of date)

File: gnss_analysis/evaluation/metrics.py (hunk 1 above)
Comment (akleeman) [inline gnss_analysis/evaluation/metrics.py]: I was picturing a situation where we'd want to pass in observations that don't necessarily have a known location (perhaps piksi logs from a kinematic test set?) but still perform a suite of not-yet-implemented evaluations, in which case we wouldn't want a hard fail.  If you don't think that's something we need to worry about then I agree hard fail is the way to go.

File: gnss_analysis/filters/common.py (hunk 5 above)
Comment (akleeman) [inline gnss_analysis/filters/common.py]: Yeah warning would be better.  Though I'm curious why `logging.info` wasn't being printed to stdout.

File: gnss_analysis/evaluation/metrics.py (hunk 2 above)
Comment (akleeman) [inline gnss_analysis/evaluation/metrics.py]: done

File: gnss_analysis/filters/__init__.py (hunk 4 above)
Comment (akleeman) [inline gnss_analysis/filters/__init__.py]: done

File: gnss_analysis/tools/compare.py (hunk 6 above)
Comment (akleeman) [inline gnss_analysis/tools/compare.py]: done

File: gnss_analysis/evaluation/metrics.py (hunk 1 above)
Comment (benjamin0) [inline gnss_analysis/evaluation/metrics.py]: sounds good!

**Comment by benjamin0** [comment]
lgtm!

**Pull request merged**